### DISCUSSION TITLE
Getters and setters

### DISCUSSION THREAD

**Discussion by clmnin** [geldata/gel#3013] [Ideas]
We've usually used an `int` field in a DB when storing a currency value, mostly because currency only need a precision of "upto two decimal places".

The schema for a currency property (price) could be an `int32` in edgedb

```sql
module default {
    type Product {
        required property name -> str,
        required property price -> int32,
    }
}
```

But before we INSERT or SELECT the data, we have to either multiple by 100 or divide by 100 (in the specific currency usecase). If this could be defined as getters and setters in the schema that would be an amazing feature.

This is how Kotlin does it
```kotlin
var someProperty: String = "defaultValue"
    get() = field
    set(value) { field = value }
```

**Proposal**
```sql
module default {
    type Product {
        required property name -> str,
        required property price -> int32 {
            get := field/100;
            set(value) := field*100;
        }
    }
}
```

**Comment by CodesInChaos** [comment]
I'd use a custom scalar type for this, plus conversion functions:

    scalar type Currency extending int32;

    function currency_to_decimal(x: Currency) -> decimal
    using (
        SELECT <decimal>x / 100
    );

    function decimal_to_currency(x: decimal) -> Currency
    using (
        SELECT <Currency>(x * 100)
    );

**Reply by elprans** [reply]
The `getter` part can be implemented as a computed property for convenience.

```
price_decimal := round(<decimal>.price / 100, 2)
```
